### PR TITLE
Simplify sockets

### PR DESCRIPTION
**Pull request by jmfrank63**
This is code to look at as part of #542 It seems static analysis tools are not able to pick up any dynamically added attributes.
The only chance if we want code completion for all the `trio.socket` attributes is to define them statically.
I have an example in socket.py that works, it looks complicated and is a leftover of my try to add the attributes to a class dynamically hoping the static analysis would pick it up.
But to no avail. https://github.com/davidhalter/jedi/issues/997

Code generation might be able to do this, but that will make it even more difficult to deal with the subject.

There is a discussion open at https://github.com/davidhalter/jedi/issues/385 which could point out the problem being solved on the other end of the stick.

@Zac-HD @njsmith What do you think?

I think this request will be closed after the discussion. The only reason I could think of it being accepted (with modification) is the micro optimisation of _reexport to update the globals once with a full dictionary instead of once per loop.

### CODE REVIEW

**Comment by jmfrank63** [comment]
I will go ahead and do a static implementation to have a usable example on which we can judge about. I am aware that the socket constants are highly volatile, but at the moment this is our only option if we want code completion to be working.

**Comment by njsmith** [comment]
I think you're making this more complicated than you need to :-)

For the first version, we want to:

* Take the re-exporting of all-caps constants that's in `trio/_socket.py`, and move it into `trio/socket.py`. (Yes, this means that tab completion won't work for these symbols. Too bad. It's probably not fixable, and to the extent it is, we can worry about it later :-).)

* Remove the explicit `__all__` management from `trio/_socket.py` and `trio/socket.py`

* For the lowercase exports, add an explicit `from ._socket import (... long list here ...)` to `trio/socket.py`

* I think that's all?

**Comment by jmfrank63** [comment]
I am getting more and more a deeper understanding of the problem. It does not seem to be feasible to do a static import, as the names are version and os dependent. With 3 versions and 3 operating systems there are already 9 possible scenarios. But since importing from socket is leading to a functional code completion and since trio.socket is meant as a drop in replacement to socket I will do a star import and implement all functions that are currently not wanted and raising an NotImplemented exception. That seems to be a much cleaner approach and will hopefully finally work.

**Comment by njsmith** [comment]
The test failure is actually a Python bug! See [bpo-34759](https://bugs.python.org/issue34759) and https://github.com/python/cpython/pull/9468. Fortunately it hadn't yet made it into any releases, but the next 3.6 and 3.7 point releases are branching *tonight*, so we really caught this in the nick of time...

**Comment by Zac-HD** [comment]
(not subscribed to this thread since know almost nothing about sockets, but @ me and I'll respond!)

> A larger question, where I'm not sure of the answer: right now the way this PR is structured, we import all the ALLCAPS constants from `socket` → `trio._socket`, and then we import them again from `trio._socket` → `trio.socket`. Maybe it would be simpler to take them out of `trio._socket`, and instead import them directly `socket` → `trio.socket`?

I would certainly find it more tasteful to import them directly from `socket` → `trio.socket`.

> What makes this complicated is that there are some constants that are missing from `socket`, so we define them ourselves. And one of those constants, `IPPROTO_IPV6`, we actually [use](https://github.com/python-trio/trio/blob/f7ccd96ac0a06e17658aa535104bf5c8edbe0f99/trio/_socket.py#L541-L543) inside `trio._socket`. So we can't *just* move all this stuff to `trio.socket`.

Define as many as possible in `trio.socket`, and define-and-export those we need in `trio._socket` I guess.  This is *probably* better than defining them twice, but I could easily be argued around.

> I guess it would still be simpler to have everything except `IPPROTO_IPV6` imported directly from `socket` → `trio.socket`, and just special-case `IPPROTO_IPV6`?

Yep 😥

File: trio/_socket.py [hunk 1]
```diff
@@ -1,39 +1,37 @@
...
+    inet_ntop,
+)
 
-def _add_to_all(obj):
-    __all__.append(obj.__name__)
-    return obj
+try:
+    from socket import (
+        if_nameindexif_nametoindex, if_indextoname, sethostname
+    )
```
[the remark this answers is not in the document]
Comment (jmfrank63) [reply trio/_socket.py]: I thought I had fixed this already but I'll take a look again. 

But this typo shows the weakness of this approach. I had to put these four into a try catch block because on py3.5 they are not there. Actually I should put each of the names in a try except block of its own but that would make the code even uglier.

So we do not only have a constant version/os dependency but also for functions.

I can see you are somehow reluctant towards a star import with overwriting which I now would prefer but I still do not know why exactly. Now you have about 25 years more experience than me so this is very likely well founded, I'd just like to understand why.

File: trio/_socket.py [hunk 2]
```diff
@@ -549,6 +550,7 @@ def shutdown(self, flag):
         # empty list.
         assert len(gai_res) >= 1
         # Address is the last item in the first entry
+        # normed = gai_res[0][-1]
```
[the remark this answers is not in the document]
Comment (jmfrank63) [reply trio/_socket.py]: I put it in because it does the same as your assignment but pylint doesn't complain. Was meant as a reminder to change this one day so we have one complaint less.
(actually this one was the last standing) But this should be very likely if at all a different PR. Right?

File: trio/_core/_io_windows.py [hunk 3]
```diff
@@ -341,8 +341,8 @@ def monitor_completion_key(self):
         if sock in self._socket_waiters[which]:
             await _core.checkpoint()
             raise _core.ResourceBusyError(
-                "another task is already waiting to {} this socket"
-                .format(which)
+                "another task is already waiting to {} this socket".
```
Comment (belm0) [inline trio/_core/_io_windows.py]: suggest that unrelated code formatting be done in a separate PR

File: trio/_socket.py [hunk 4]
```diff
@@ -310,13 +293,12 @@ def fromfd(*args, **kwargs):
 if hasattr(_stdlib_socket, "fromshare"):
 
     @_wraps(_stdlib_socket.fromshare, assigned=(), updated=())
-    @_add_to_all
+    # @_add_to_all
```
[the remark this answers is not in the document]
Comment (belm0) [reply trio/_socket.py]: @jmfrank63  in general, please be sure that new pushes are addressing outstanding comments.  Otherwise it seems that github automatically marks the comments as resolved, and they git lost.  I think this was a pervasive communication problem on your last PR.

File: trio/_socket.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (njsmith) [reply trio/_socket.py]: Oh, does pylint complain about the unpacking assignment? That's weird, it seems more readable to me...

And yes, better to separate it out in any case.

File: trio/_socket.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jmfrank63) [reply trio/_socket.py]: @belm0 Thanks for letting me know, I will go through all of them and address them in the next push. There is so much to learn, thanks for the patience. I will also try to attempt squashing this time, as there is so much back and forth.

[the remark this answers is not in the document]
Comment (belm0) [reply trio/_socket.py]: Thank you for considering.  (And again it's dubious that GitHub behaves that way.)

About squashing, that can be done trivially by the person doing the final merge into master.

File: trio/_socket.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (njsmith) [reply trio/_socket.py]: Huh, yeah, looks like a pylint bug – it's supposed to catch illegal code like `*a, *b = c`, but it also fires on valid code like this. Filed at https://github.com/PyCQA/pylint/issues/2513

File: trio/_core/_io_windows.py (hunk 3 above)
Comment (jmfrank63) [reply trio/_core/_io_windows.py]: @belm0 Oh, I didn't touch that module at all, guess that came from yapf and it did slip in with a `git add --all` which I learned I shouldn't do. I'll have to check how to revert a single file of a commit to redo this change, I guess that should be possible.

File: trio/socket.py [hunk 5]
```diff
@@ -1,8 +1,43 @@
...
-from ._socket import __all__
+import importlib
+
+_smod = importlib.import_module('._socket', 'trio')
+
+from ._socket import (
+    fromfd, from_stdlib_socket, getprotobyname, socketpair, getnameinfo,
+    socket, getaddrinfo, set_custom_hostname_resolver,
+    set_custom_socket_factory, real_socket_type, SocketType, fspath,
+    run_sync_in_worker_thread
```
[the remark this answers is not in the document]
Comment (jmfrank63) [reply trio/socket.py]: Removed. Shouldn't be in the next push.

File: trio/socket.py [hunk 6]
```diff
@@ -1,8 +1,43 @@
...
+# attributes that arent actually part of our public API. But its very
 # annoying to carefully always use underscored names for module-level
 # temporaries, imports, etc. when implementing the module. So we put the
 # implementation in an underscored module, and then re-export the public parts
 # here.
-from ._socket import *
-from ._socket import __all__
+import importlib
+
+_smod = importlib.import_module('._socket', 'trio')
```
[the remark this answers is not in the document]
Comment (jmfrank63) [reply trio/socket.py]: Yes, done. I tried to pull directly into socket.py whatever I could. Interestingly the python interactive shell with jedi configured does have full code completion, while ipython hangs. Either some implications of pip install -e or just bugs.
Anyways it currently looks good.

File: trio/socket.py [hunk 7]
```diff
@@ -1,8 +1,43 @@
-# This is a public namespace, so we don't want to expose any non-underscored
-# attributes that aren't actually part of our public API. But it's very
+# This is a public namespace, so we dont want to expose any non-underscored
+# attributes that arent actually part of our public API. But its very
```
[the remark this answers is not in the document]
Comment (jmfrank63) [reply trio/socket.py]: Back in, deleted by accident when doing a replace all.

File: trio/_socket.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jmfrank63) [reply trio/_socket.py]: Reported 16 hours ago, fixed an hour ago, how do you do this? 
Besides I would have never considered this a bug in pylint. So amazing.
Anyways, removed the comment again.

File: trio/_socket.py [hunk 8]
```diff
@@ -62,54 +60,46 @@ def _is_blocking_io_error(self, exc):
...
+try:
+    from socket import IPPROTO_IPV6
+except ImportError:
     # As of at least 3.6, python on Windows is missing IPPROTO_IPV6
     # https://bugs.python.org/issue29515
-    if not hasattr(_stdlib_socket, "IPPROTO_IPV6"):  # pragma: no branch
+    if _sys.platform == "win32":  # pragma: no branch
         IPPROTO_IPV6 = 41
-        __all__.append("IPPROTO_IPV6")
+        globals()['IPPROTO_IPV6'] = IPPROTO_IPV6
```
[the remark this answers is not in the document]
Comment (jmfrank63) [reply trio/_socket.py]: Removed the noop statements.

File: trio/_socket.py [hunk 9]
```diff
@@ -62,54 +60,46 @@ def _is_blocking_io_error(self, exc):
...
-    # See https://github.com/python-trio/trio/issues/39
-    # (you can still get it from stdlib socket, of course, if you want it)
-    globals().pop("SO_REUSEADDR", None)
-    __all__.remove("SO_REUSEADDR")
+try:
+    from socket import IPPROTO_SCTP
+except ImportError:
+    if _sys.platform == "darwin":
+        IPPROTO_SCTP = 132
+        globals()['IPPROTO_SCTP'] = IPPROTO_SCTP
```
[the remark this answers is not in the document]
Comment (jmfrank63) [reply trio/_socket.py]: Removed all unnecessary definitions from _socket.py. I did also just remove the import of SO_REUSEADDR from _socket.py as it is not used and deleted it in socket.py.

File: trio/_socket.py [hunk 10]
```diff
@@ -1,39 +1,37 @@
...
 
 
-# Usage:
-#
-#   async with _try_sync():
-#       return sync_call_that_might_fail_with_exception()
-#   # we only get here if the sync call in fact did fail with a
-#   # BlockingIOError
-#   return await do_it_properly_with_a_check_point()
-#
```
[the remark this answers is not in the document]
Comment (jmfrank63) [reply trio/_socket.py]: Comment reintroduced deletion was an accident, held it for my own. I never touch comment of others.

File: trio/_socket.py [hunk 11]
```diff
@@ -1,39 +1,37 @@
 import os as _os
-import socket as _stdlib_socket
 import sys as _sys
+import importlib as _importlib
 from functools import wraps as _wraps
 
+_stdlib_socket = _importlib.import_module('socket')
```
[the remark this answers is not in the document]
Comment (jmfrank63) [reply trio/_socket.py]: Any use of importlib removed again, as there is no benefit and it makes the code only harder to read.

**Comment by jmfrank63** [comment]
Hmmh, didn't touch any of the tests, and in theory this message should be filtered out at line 596.

File: trio/socket.py [hunk 12]
```diff
@@ -4,5 +4,71 @@
...
+
+from . import _socket
+import sys as _sys
+
+# import the overwrites
+from ._socket import (
+    fromfd, from_stdlib_socket, getprotobyname, socketpair, getnameinfo,
+    socket, getaddrinfo, set_custom_hostname_resolver,
+    set_custom_socket_factory, SocketType, _stdlib_socket
+)
```
[the remark this answers is not in the document]
Comment (jmfrank63) [reply trio/socket.py]: Yep, nicer, so all our own defined stuff is in one place.

File: trio/socket.py [hunk 13]
```diff
@@ -4,5 +4,71 @@
...
+
+if _sys.platform == 'win32':
+    # See https://github.com/python-trio/trio/issues/39
+    # Do not import for windows platform
+    # (you can still get it from stdlib socket, of course, if you want it)
+    try:
+        del SO_REUSEADDR
+    except NameError:
+        pass
```
[the remark this answers is not in the document]
Comment (jmfrank63) [reply trio/socket.py]: Done.

File: trio/socket.py [hunk 14]
```diff
@@ -4,5 +4,71 @@
...
+
+# get names used by trio that might not be available in all versions
+try:
+    IPPROTO_IPV6
+except NameError:
+    from ._socket import IPPROTO_IPV6
```
[the remark this answers is not in the document]
Comment (jmfrank63) [reply trio/socket.py]: try/except removed, let's see if we still pass...

File: trio/socket.py [hunk 15]
```diff
@@ -4,5 +4,71 @@
...
+
+try:
+    TCP_NOTSENT_LOWAT
+except NameError:
+    if _sys.platform != 'win32':
+        from ._socket import TCP_NOTSENT_LOWAT
```
[the remark this answers is not in the document]
Comment (jmfrank63) [reply trio/socket.py]: Done.

File: trio/_socket.py [hunk 16]
```diff
@@ -62,54 +46,13 @@ def _is_blocking_io_error(self, exc):
...
-    globals().pop("SO_REUSEADDR", None)
-    __all__.remove("SO_REUSEADDR")
-
+try:
+    from socket import IPPROTO_IPV6
+except ImportError:
     # As of at least 3.6, python on Windows is missing IPPROTO_IPV6
     # https://bugs.python.org/issue29515
-    if not hasattr(_stdlib_socket, "IPPROTO_IPV6"):  # pragma: no branch
+    if _sys.platform == "win32":  # pragma: no branch
```
Comment (pquentin) [inline trio/_socket.py]: According to https://codecov.io/gh/python-trio/trio/pull/670/changes you don't need `pragma: no branch`: this is covered by the Windows tests.

Comment (jmfrank63) [reply trio/_socket.py]: @pquentin I removed the comment. How do I learn about codecov? What would be some good resources to read?

Comment (pquentin) [reply trio/_socket.py]: @jmfrank63 Nothing that I know of, sorry :(

File: newsfragments/670.feature.rst [hunk 17]
```diff
@@ -0,0 +1 @@
+* Reworked :mod:`trio.socket` namespace construction, making it more understandable by static analysis tools. This should improve tab completion in editors, reduce false positives from pylint, and is another step towards providing type hints.
```
Comment (njsmith) [inline newsfragments/670.feature.rst]: Thinking about it, it's going to be kind of weird-looking in our final [change log](https://trio.readthedocs.io/en/latest/history.html) to have multiple almost-identical entries with just the module name changed.

Since #542 is already an umbrella bug for these changes, how about we drop this file, and instead modify the existing `542.feature.rst` to cover the new work as well? And while we're at it, we can add mention of the work in #656 to fix `trio.testing`... so it can say something like:

"Reworked :mod:`trio`, :mod:`trio.testing`, and :mod:`trio.socket` namespace construction, making them more understandable..."

Also BTW, newsfragments shouldn't start with a `*`. Towncrier will automatically add a `*` when it assembles the release notes, and then things will look weird if we have two `*`s :-)

File: trio/_core/_run.py [hunk 18]
```diff
@@ -858,8 +858,8 @@ def _return_value_looks_like_wrong_library(value):
                     "trio was expecting an async function, but instead it got "
                     "{!r} – are you trying to use a library written for "
                     "asyncio/twisted/tornado or similar? That won't work "
-                    "without some sort of compatibility shim."
-                    .format(async_fn)
+                    "without some sort of compatibility shim.".
+                    format(async_fn)
```
Comment (njsmith) [inline trio/_core/_run.py]: I upgraded yapf (#694), so I think you need to merge from master and then re-run yapf (using 0.24.0) before we can merge. Hopefully that will fix these weird formatting issues too...

Comment (njsmith) [reply trio/_core/_run.py]: Yeah, you definitely need yapf 0.24.0

**Comment by njsmith** [comment]
Well, I filed https://github.com/google/yapf/issues/624 anyway for this annoying thing in `trio/_signals.py`.

**Comment by jmfrank63** [comment]
Looks like we are fine now, any fine tuning possibly?

**Comment by njsmith** [comment]
@jmfrank63 https://github.com/python-trio/trio/pull/670#discussion_r221138923

File: newsfragments/670.feature.rst (hunk 17 above)
Comment (jmfrank63) [reply newsfragments/670.feature.rst]: Yes, makes sense. I'll correct this later this day, currently commuting. I actually put this in in haste to have a change, before I learned that I could actually do an empty commit.

**Comment by jmfrank63** [comment]
Good to see checks are finally passing ;-) I guess we're getting close now.

File: newsfragments/542.feature.rst [hunk 19]
```diff
@@ -1 +1 @@
-* Reworked :mod:`trio` namespace construction, making it more understandable by static analysis tools. This should improve tab completion in editors, reduce false positives from pylint, and is a first step towards providing type hints.
\ No newline at end of file
+Reworked :mod:trio, :mod:trio.testing, and :mod:trio.socket namespace construction, making them more understandable by static analysis tools. This should improve tab completion in editors, reduce false positives from pylint, and is a first step towards providing type hints.
```
Comment (njsmith) [inline newsfragments/542.feature.rst]: You need the backticks around the module names :-) It's a restructured text thing: http://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#roles

```
Reworked :mod:`trio`, :mod:`trio.testing`, and :mod:`trio.socket` ...
```

**Comment by njsmith** [comment]
:tada: :balloon:
